From 4de54f598b66cd8d8fc55328f99f3b568cc9d4eb Mon Sep 17 00:00:00 2001
From: Hans de Goede <hdegoede@redhat.com>
Date: Tue, 17 Jul 2018 09:46:12 +0200
Subject: [PATCH 1/3] logger: Add a separator between different boot logs

Since we concatenate boot logs one after the other in /var/log/boot.log
it is hard to tell where the logs from one boot end the next boot starts.

This commit makes plymouth write out a separator including a time + date
of the date, when the log file gets opened to add new boot messages to it.

Note ply_logger_open_file() is only called from ply_terminal_session_open_log()
which in turn is only used for /var/log/boot.log, so this only effects
/var/log/boot.log.

Closes #29

Signed-off-by: Hans de Goede <hdegoede@redhat.com>
---
 src/libply-splash-core/ply-device-manager.c |   16 +++++++++++-----
 src/libply/ply-logger.c                     |   13 +++++++++++++
 src/plugins/splash/two-step/plugin.c        |    6 ++++--
 3 files changed, 28 insertions(+), 7 deletions(-)

Index: b/src/libply/ply-logger.c
===================================================================
--- a/src/libply/ply-logger.c
+++ b/src/libply/ply-logger.c
@@ -34,6 +34,7 @@
 #include <sys/file.h>
 #include <sys/stat.h>
 #include <sys/types.h>
+#include <time.h>
 #include <unistd.h>
 
 #include "ply-utils.h"
@@ -312,6 +313,9 @@ bool
 ply_logger_open_file (ply_logger_t *logger,
                       const char   *filename)
 {
+        char header[80];
+        struct tm* tm;
+        time_t t;
         int fd;
 
         assert (logger != NULL);
@@ -328,6 +332,15 @@ ply_logger_open_file (ply_logger_t *logg
 
         logger->filename = strdup (filename);
 
+        time (&t);
+        tm = localtime (&t);
+        if (tm) {
+                /* This uses uname -v date format */
+                strftime (header, sizeof(header),
+                         "------------ %a %b %d %T %Z %Y ------------\n", tm);
+                ply_logger_write (logger, header, strlen(header), true);
+        }
+
         return true;
 }
 
Index: b/src/libply-splash-core/ply-device-manager.c
===================================================================
--- a/src/libply-splash-core/ply-device-manager.c
+++ b/src/libply-splash-core/ply-device-manager.c
@@ -81,6 +81,8 @@ struct _ply_device_manager
 
         uint32_t                    paused : 1;
         uint32_t                    device_timeout_elapsed : 1;
+        uint32_t                    found_drm_device : 1;
+        uint32_t                    found_fb_device : 1;
 };
 
 static void
@@ -283,6 +285,12 @@ create_devices_for_udev_device (ply_devi
                                                                                  device_path,
                                                                                  terminal,
                                                                                  renderer_type);
+                        if (created) {
+                                if (renderer_type == PLY_RENDERER_TYPE_DRM)
+                                        manager->found_drm_device = 1;
+                                if (renderer_type == PLY_RENDERER_TYPE_FRAME_BUFFER)
+                                        manager->found_fb_device = 1;
+                        }
                 }
         }
 
@@ -847,8 +855,6 @@ create_non_graphical_devices (ply_device
 static void
 create_devices_from_udev (ply_device_manager_t *manager)
 {
-        bool found_drm_device, found_fb_device;
-
         manager->device_timeout_elapsed = true;
 
         if (manager->paused) {
@@ -858,10 +864,10 @@ create_devices_from_udev (ply_device_man
 
         ply_trace ("Timeout elapsed, looking for devices from udev");
 
-        found_drm_device = create_devices_for_subsystem (manager, SUBSYSTEM_DRM);
-        found_fb_device = create_devices_for_subsystem (manager, SUBSYSTEM_FRAME_BUFFER);
+        create_devices_for_subsystem (manager, SUBSYSTEM_DRM);
+        create_devices_for_subsystem (manager, SUBSYSTEM_FRAME_BUFFER);
 
-        if (found_drm_device || found_fb_device)
+        if (manager->found_drm_device || manager->found_fb_device)
                 return;
 
         ply_trace ("Creating non-graphical devices, since there's no suitable graphics hardware");
Index: b/src/plugins/splash/two-step/plugin.c
===================================================================
--- a/src/plugins/splash/two-step/plugin.c
+++ b/src/plugins/splash/two-step/plugin.c
@@ -1002,10 +1002,12 @@ add_pixel_display (ply_boot_splash_plugi
                                             (ply_pixel_display_draw_handler_t)
                                             on_draw, view);
         if (plugin->is_visible) {
-                if (view_load (view))
+                if (view_load (view)) {
                         ply_list_append_data (plugin->views, view);
-                else
+                        view_start_progress_animation (view);
+                } else {
                         view_free (view);
+                }
         } else {
                 ply_list_append_data (plugin->views, view);
         }
